src/game/logic/object/weaponset.cpp

```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
+        if ((mask & m_totalAntiMask) == 0) {
+            return ATTACK_RESULT_1;
+        }
+
+        if (victim_obj == nullptr) {
+            return result;
+        }
+
+        if (!Is_Any_Within_Target_Pitch(source_obj, victim_obj)) {
+            return ATTACK_RESULT_1;
+        }
+
+        WeaponSlotType start_slot;
+        WeaponSlotType end_slot;
+
+        if (Is_Cur_Weapon_Locked()) {
+            start_slot = m_curWeapon;
+            end_slot = m_curWeapon;
+        } else if (wslot == WEAPONSLOT_UNK) {
+            start_slot = WEAPONSLOT_TERTIARY;
+            end_slot = WEAPONSLOT_PRIMARY;
+        } else {
+            start_slot = wslot;
+            end_slot = wslot;
+        }
+
+        for (int i = start_slot; i >= end_slot; i--) {
+            Weapon *weapon2 = m_weapons[i];
+
+            if (weapon2 != nullptr && weapon2->Estimate_Weapon_Damage(source_obj, victim_obj) != 0.0f
+                && (weapon2->Get_Damage_Type() != DAMAGE_KILL_PILOT || !source_obj->Is_KindOf(KINDOF_HERO)
+                    || m_curWeapon != WEAPONSLOT_PRIMARY || wslot != WEAPONSLOT_UNK)) {
+                return result;
+            }
+        }
+    }
+
+    ContainModuleInterface *contain = source_obj->Get_Contain();
+
+    if (contain != nullptr) {
+        if (contain->Is_Passenger_Allowed_To_Fire(OBJECT_UNK)) {
+            const std::list<Object *> *list = contain->Get_Contained_Items_List();
+
+            if (list != nullptr) {
+                for (auto it = list->begin(); it != list->end(); it++) {
+                    Object *obj = *it;
+
+                    if (obj->Is_Able_To_Attack()) {
+                        CanAttackResult can =
+                            obj->Get_Able_To_Use_Weapon_Against_Target(type, victim_obj, location, source, WEAPONSLOT_UNK);
+
+                        if (can == ATTACK_RESULT_3 || can == ATTACK_RESULT_2) {
+                            return can;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    SpawnBehaviorInterface *behavior = source_obj->Get_Spawn_Behavior_Interface();
+
+    if (behavior == nullptr
+        || behavior->Get_Can_Any_Slaves_Use_Weapon_Against_Target(type, victim_obj, location, source) != ATTACK_RESULT_3) {
+        return ATTACK_RESULT_1;
+    }
+
+    if (source_obj->Is_KindOf(KINDOF_IMMOBILE) && source_obj->Is_KindOf(KINDOF_SPAWNS_ARE_THE_WEAPONS)
+        && result == ATTACK_RESULT_2) {
+        return ATTACK_RESULT_3;
+    }
+
+    return result;
+}
+
+bool WeaponSet::Choose_Best_Weapon_For_Target(
+    const Object *source_obj, const Object *victim_obj, WeaponChoiceCriteria criteria, CommandSourceType source)
+{
+    if (Is_Cur_Weapon_Locked()) {
+        return true;
+    }
+
+    if (victim_obj != nullptr) {
+        bool b1 = false;
+        bool b2 = false;
+        float f1 = 0.0f;
+        float f2 = 0.0f;
+        float f3 = 0.0f;
+        float f4 = 0.0f;
+        int i1 = 0;
+        int i2 = 0;
```

Comment:
free_weapon
busy_weapon

Can be `WeaponSlotType`

Reply:
fixed.